File: HeraldTest-Flutter/ios/Runner/IllnessStatusPayloadDataSupplier.swift
//
//  IllnessStatusPayloadDataSupplier.swift
//  Runner
//
//  Created by Andi Grozdani on 11/2/21.
//

//
//  IllnessStatusPayloadDataSupplier.swift
//  HeraldTest
//
//

import Foundation
import Herald

class IllnessDataPayloadSupplier: PayloadDataSupplier {
    
    fileprivate var identifier:Int;
    fileprivate var status: IllnessStatus; // set by only constructor

    init(identifier: Int, status: IllnessStatus) {
        self.identifier = identifier
        self.status = status
    }

    init(identifier: Int) {
        self.identifier = identifier
        status = IllnessStatus(status: .susceptible, dateSince: Date())
    }

    func setStatus(newStatus: IllnessStatus) {
        self.status = newStatus
    }

    func getStatus() -> IllnessStatus {
        return status
    }

    func getIdentifier() -> Int{
        return identifier
    }

    func setIdentifier(identifier: Int) {
        self.identifier = identifier
    }

    func payload(_ timestamp: PayloadTimestamp, device: Device?) -> PayloadData? {
        // Return our status as a series of Bytes in a Data object instance
        // Note: Should cache this if it doesn't change often as a member in order to save processing CPU and thus battery
        let result = PayloadData();
        result.append(UInt64(identifier)); // bytes 0-7
        result.append(status.toPayload()); // bytes 8-16
        return result;
    }

    static func getIdentifierFromPayload(illnessPayload: PayloadData) -> Int {
        return Int(illnessPayload.data.uint64(0)!); // use first 8 bytes only
    }

    static func getIllnessStatusFromPayload(illnessPayload: PayloadData) -> IllnessStatus {
        return IllnessStatus.fromPayload(raw: illnessPayload.data.advanced(by: 8)); // don't include first 8 bytes
    }
    
}

File: HeraldTest-Flutter/android/app/src/main/java/com/example/flutterxherald/IllnessStatusPayloadDataSupplier.java
package com.example.flutterxherald;

import io.heraldprox.herald.sensor.datatype.Date;
import io.heraldprox.herald.sensor.datatype.UInt64;
import io.heraldprox.herald.sensor.datatype.PayloadData;
import io.heraldprox.herald.sensor.datatype.PayloadTimestamp;
import io.heraldprox.herald.sensor.Device;
import io.heraldprox.herald.sensor.payload.DefaultPayloadDataSupplier;

public class IllnessStatusPayloadDataSupplier extends DefaultPayloadDataSupplier {
    protected int identifier;
    protected IllnessStatus status; // set by only constructor

    public IllnessStatusPayloadDataSupplier(int identifier,IllnessStatus initialStatus) {
        this.identifier = identifier;
        status = initialStatus;
    }

    public IllnessStatusPayloadDataSupplier(int identifier) {
        this.identifier = identifier;
        status = new IllnessStatus(IllnessStatusCode.susceptible,new Date());
    }

    public void setStatus(IllnessStatus newStatus) {
        this.status = newStatus;
    }

    public IllnessStatus getStatus() {
        return status;
    }

    public int getIdentifier() {
        return identifier;
    }

    public void setIdentifier(int identifier) {
        this.identifier = identifier;
    }

    /**
     * Returns a payload to be sent by Herald to nearby devices upon request (Payload Read)
     * @param timestamp The timestamp for which the read was requested. May be null.
     * @param device The virtual nearby Herald device reference (unique per remote requestor). May be null.
     * @return PayloadData instance containing UInt64 dateSince and UInt8 illnessStatus
     */
    public PayloadData payload(@androidx.annotation.NonNull PayloadTimestamp timestamp,
                               @androidx.annotation.Nullable Device device) {
        // Return our status as a series of Bytes in a Data object instance
        // Note: Should cache this if it doesn't change often as a member in order to save processing CPU and thus battery
        PayloadData result = new PayloadData();
        result.append(new UInt64(identifier)); // bytes 0-7
        result.append(status.toPayload()); // bytes 8-16
        return result;
    }

    static int getIdentifierFromPayload(PayloadData illnessPayload) {
        return (int)illnessPayload.uint64(0).value; // use first 8 bytes only
    }

    static IllnessStatus getIllnessStatusFromPayload(PayloadData illnessPayload) {
        return IllnessStatus.fromPayload(illnessPayload.subdata(8)); // don't include first 8 bytes
    }
}
